File: ProgettoMobidev/app/src/main/java/com/capraraedefrancescosoft/progettomobidev/models/Journals.java
package com.capraraedefrancescosoft.progettomobidev.models;

import android.util.Log;

import com.google.firebase.messaging.FirebaseMessaging;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Ianfire on 03/07/2016.
 */
public class Journals {

    private static Journals instance = new Journals();

    private Map<JournalID, Journal> journals;
    private JournalID currentJournalID;

    private Journals() {
        this.journals = new HashMap<>();
        this.currentJournalID = null;
    }

    public static Journals getInstance() {
        return instance;
    }

    public synchronized void setJournals(List<Journal> journals) {
        this.journals = new HashMap<>();
        // per ogni journal parte la sottoiscrizione al topic del journal
        // e viene aggiunto all'hashmap
        for (Journal journal : journals){
            this.journals.put(new JournalID(journal.getOwnerID(), journal.getName()), journal);

            String topic = journal.getOwnerID() + "_" + journal.getName().replace(" ", "_");
            FirebaseMessaging.getInstance().subscribeToTopic(topic);
            System.out.println("Sottoscritto al topic " + topic);
        }
    }

    public synchronized JournalID addJournal(Journal journal){
        JournalID id = new JournalID(journal.getOwnerID(), journal.getName());
        journals.put(id, journal);

        return id;
    }

    /** Ritorna il journal cercato */
    public synchronized Journal getJournal(JournalID id){
        return journals.get(id);
    }

    public List<Journal> getJournals(){
        return new ArrayList<Journal>(journals.values());
    }

    public synchronized void setCurrentJournal(JournalID id) {
        this.currentJournalID = id;
    }

    public synchronized Journal getCurrentJournal() {
        return journals.get(currentJournalID);
    }

    /** Rimpiazza il journal con quello nuovo fornito */
    public JournalID updateJournal(Journal newJournal) {
        JournalID newID = new JournalID(newJournal.getOwnerID(), newJournal.getName());
        Journal precedente = journals.put(newID, newJournal);

        if(precedente == null){
            Log.e("TravelNotes", "Errore: updateJournal non ha trovato il journal che doveva rimpiazzare");
        }

        return newID;
    }

}

File: ProgettoMobidev_iOS/progettomobidev/Journals.m
//
//  Journals.m
//  Travel Notes
//
//  Created by gianpaolo on 07/10/16.
//  Copyright © 2016 capraraedefrancescosoft. All rights reserved.
//

#import "Journals.h"

@implementation Journals

@synthesize journals,currentJournalID;

#pragma mark - singleton method

+ (Journals*)getInstance
{
    static Journals *sharedJournals = nil;
    @synchronized(self) {
        if (sharedJournals == nil)
            sharedJournals = [[self alloc] init];
    }
    return sharedJournals;
}

- (Journals*)init{
    if (self = [super init]){
        journals = [NSMutableDictionary dictionary];
        currentJournalID = [[JournalID alloc] init];
    }
    return self;
}

- (void)dealloc {
    // Should never be called, but just here for clarity really.
}

- (void) setJournalsInDict: (NSMutableArray*) newJournals{
    journals = [NSMutableDictionary dictionary];
    // per ogni journal parte la sottoiscrizione al topic del journal
    // e viene aggiunto all'hashmap
    for (Journal* journal in newJournals){
        [journals setObject:journal forKey: (id <NSCopying>)[[JournalID alloc] initWithName:journal.name andOwnerId:journal.owner_id]];
    }
}

- (JournalID*) addJournal: (Journal*) journal{
    JournalID* newID = [[JournalID alloc] initWithName:journal.name andOwnerId:journal.owner_id];
    [journals setObject:journal forKey: (id <NSCopying>)newID];
    
    return newID;
}

- (Journal*) getJournal: (JournalID*) journal_id{
    return [journals objectForKey:(id <NSCopying>)journal_id];
}

- (NSMutableArray*) getJournals{
    NSMutableArray* returnArray = [NSMutableArray arrayWithArray: [journals allValues]];
    return returnArray;
}

- (void) setCurrentJournal: (JournalID*) journal_id{
    currentJournalID = journal_id;
}

- (Journal*) getCurrentJournal{
    return [journals objectForKey:(id <NSCopying>)currentJournalID];
}

- (JournalID*) updateJournal: (Journal*) journal{
    JournalID* newID = [[JournalID alloc] initWithName:journal.name andOwnerId:journal.owner_id];
    [journals setObject:journal forKey: (id <NSCopying>)newID];
    /*Journal precedente = journals.put(newID, newJournal);
     
     if(precedente == null){
     Log.e("TravelNotes", "Errore: updateJournal non ha trovato il journal che doveva rimpiazzare");
     }*/
    
    return newID;
}

@end
